### ISSUE TITLE
no-extra-parens false positive for TypeScript cast

### ISSUE THREAD

**Issue by awvalenti** [eslint/eslint#13159]
**Tell us about your environment**

* **ESLint Version:** 6.8.0
* **Node Version:** 11.10.0
* **npm Version:** 6.14.4

**What parser (default, Babel-ESLint, etc.) are you using?** @typescript-eslint/parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yml
env:
  browser: true
  es6: true
  mocha: true
  node: true
extends:
  - eslint:recommended
  - plugin:react/recommended
  - plugin:@typescript-eslint/recommended
globals:
  module: readonly
  require: readonly
parser: '@typescript-eslint/parser'
parserOptions:
  ecmaFeatures:
    jsx: true
  ecmaVersion: 2020
  sourceType: module
plugins:
  - '@typescript-eslint'
  - react
  - react-hooks
rules:
  '@typescript-eslint/explicit-function-return-type': 'off'
  '@typescript-eslint/member-delimiter-style':
    - error
    - singleline:
        delimiter: semi
      multiline:
        delimiter: none
  '@typescript-eslint/no-explicit-any': 'off'
  '@typescript-eslint/no-non-null-assertion': 'off'
  '@typescript-eslint/no-unused-vars':
    - error
    - argsIgnorePattern: ^_
  array-bracket-spacing: error
  arrow-parens:
    - error
    - as-needed
  arrow-spacing:
    - error
    - after: true
      before: true
  block-scoped-var: error
  brace-style: error
  comma-dangle:
    - error
    - arrays: always-multiline
      objects: always-multiline
      imports: always-multiline
      exports: always-multiline
      functions: only-multiline
  comma-spacing: error
  comma-style: error
  complexity:
    - error
    - 4
  consistent-return: error
  default-case: error
  dot-notation: error
  eol-last:
    - error
    - always
  eqeqeq: error
  func-names:
    - error
    - never
  indent:
    - error
    - 2
    - SwitchCase: 1
  key-spacing: error
  keyword-spacing: error
  linebreak-style:
    - error
    - unix
  max-depth: warn
  max-nested-callbacks:
    - warn
    - 6
  new-cap: error
  new-parens: error
  no-alert: error
  no-array-constructor: error
  no-console: warn
  no-dupe-class-members: 'off'
  no-duplicate-imports:
    - error
    - includeExports: true
  no-eq-null: error
  no-eval: error
  no-extra-bind: error
  no-extra-label: error
  no-extra-parens: error
  no-floating-decimal: error
  no-implicit-coercion: error
  no-iterator: error
  no-lone-blocks: error
  no-lonely-if: error
  no-loop-func: error
  no-multi-spaces: error
  no-multi-str: error
  no-multiple-empty-lines:
    - error
    - max: 1
      maxBOF: 0
      maxEOF: 0
  no-negated-condition: error
  no-new: error
  no-new-func: error
  no-new-object: error
  no-octal-escape: error
  no-param-reassign: error
  no-proto: error
  no-restricted-syntax:
    - error
    - WithStatement
  no-return-assign:
    - error
    - always
  no-self-compare: error
  no-sequences: error
  no-shadow-restricted-names: error
  no-template-curly-in-string: error
  no-throw-literal: error
  no-trailing-spaces: error
  no-undef:
    - error
    - typeof: true
  no-undef-init: error
  no-unmodified-loop-condition: error
  no-unneeded-ternary: error
  no-unused-expressions: error
  no-use-before-define:
    - error
    - nofunc
  no-useless-call: error
  no-useless-computed-key: error
  no-useless-concat: error
  no-useless-constructor: error
  no-useless-rename: error
  no-void: error
  no-whitespace-before-property: error
  no-with: error
  object-curly-spacing:
    - error
    - always
  object-shorthand: error
  operator-assignment: error
  operator-linebreak: error
  quote-props:
    - error
    - consistent-as-needed
  quotes:
    - error
    - single
    - avoidEscape: true
  radix: error
  react-hooks/exhaustive-deps: warn
  react-hooks/rules-of-hooks: error
  react/prop-types: off
  rest-spread-spacing: error
  semi:
    - error
    - never
  semi-spacing:
    - error
    - after: true
      before: false
  space-before-blocks: error
  space-before-function-paren:
    - error
    - never
  space-in-parens: error
  space-infix-ops: error
  space-unary-ops: error
  spaced-comment:
    - error
    - always
    - block:
        balanced: true
        markers:
          - '!'
      line:
        exceptions:
          - '-'
  strict:
    - error
    - safe
  symbol-description: error
  template-curly-spacing: error
  unicode-bom: error
settings:
  import/resolver:
    node:
      extensions:
        - .js
        - .jsx
  react:
    version: detect

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
let a: number | string = 'text'
if (Math.random() < 0.5) a = 1
// eslint-disable-next-line no-console
console.log((a as string).toLowerCase())
```

```bash
npx eslint parens.ts
```

**What did you expect to happen?**
No output from eslint. no-extra-parens should give no error.

**What actually happened? Please include the actual, raw output from ESLint.**
```
  4:13  error  Unnecessary parentheses around expression  no-extra-parens
```

If I run it using `--fix`, it removes the (actually necessary) parens and outputs:
```
  4:35  error  Parsing error: ',' expected
```

**Are you willing to submit a pull request to fix this bug?**
Yes, if I get some help getting started in the project

**Comment by anikethsaha** [comment]
Can you turn off the base rule `no-extra-parens` (eslint) and turn on the [`@typescript-eslint/no-extra-parens`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-parens.md) .
I tested the rule with `@typescript-eslint/no-extra-parens` with your example, its working fine and its lint free

**Comment by mdjermanovic** [comment]
Thanks for the issue!

As @anikethsaha explained, for TS code you should use [`@typescript-eslint/no-extra-parens`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extra-parens.md) rule from [ESLint Plugin TypeScript](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin), instead of the core `no-extra-parens` rule.

**Comment by awvalenti** [comment]
That's great! Problem solved!

As a side-note, typescript-eslint goes the opposite direction and allows things like `console.log((((a as string))).toLowerCase())`. I'll open an issue on their repo.

Thanks!
